### PR TITLE
Let Jenkins wait for deployment step is finished or timedout

### PR DESCRIPTION
**Pull request by sneufeind**
At the moment the Jenkins-Marathon-Plugin just fires and forgets deployments. Within this PR the Marathon-Deployment-API is asked afterwards if the deployment with a given id could be finished or not in time. Otherwise the Jenkins build step will fail.
To increase testability I refactored the mighty MarathonBuilder a little and put the update part into a separate class.

### CODE REVIEW

**Comment by joerg84** [comment]
@colin-msphere could you take a look?

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 1]
```diff
@@ -56,33 +56,35 @@
     private       String              filename;
     private       String              credentialsId;
     private       boolean             forceUpdate;
+    private long timeout;
```
Comment (joerg84) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: formatting

**Comment by colin-msphere** [comment]
@mesosphere-ci ok to test

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 2]
```diff
@@ -56,33 +56,35 @@
...
     }
 
     public String getAppid() {
-        return appid;
+        return this.appid;
     }
 
     @DataBoundSetter
-    public void setAppid(@Nonnull final String appid) {
+    public void setAppid(@Nonnull String appid) {
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 3]
```diff
@@ -56,33 +56,35 @@
...
         this.appid = appid;
     }
 
     public String getFilename() {
-        return filename;
+        return this.filename;
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 4]
```diff
@@ -56,33 +56,35 @@
...
     }
 
     @DataBoundSetter
-    public void setFilename(@Nonnull final String filename) {
-        if (filename.trim().length() > 0)
+    public void setFilename(@Nonnull String filename) {
+        if (filename.trim().length() > 0) {
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 5]
```diff
@@ -108,21 +110,21 @@ public boolean needsToRunAfterFinalized() {
      * @param logger a build's logger
      * @param text   message to log
      */
-    private void log(final PrintStream logger, final String text) {
+    private void log(PrintStream logger, String text) {
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 6]
```diff
@@ -108,21 +110,21 @@ public boolean needsToRunAfterFinalized() {
...
 
     @Override
     public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener)
             throws InterruptedException, IOException {
-        final boolean     buildSucceed = build.getResult() == null || build.getResult() == Result.SUCCESS;
-        final EnvVars     envVars      = build.getEnvironment(listener);
-        final PrintStream logger       = listener.getLogger();
+        boolean buildSucceed = build.getResult() == null || build.getResult() == Result.SUCCESS;
+        EnvVars envVars = build.getEnvironment(listener);
+        PrintStream logger = listener.getLogger();
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: there are aspects of this change that I really like... the spacing removal makes sense... I'm not sure why they were like that.   It is not necessary to remove the `final`

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 7]
```diff
@@ -108,21 +110,21 @@ public boolean needsToRunAfterFinalized() {
...
         envVars.overrideAll(build.getBuildVariables());
 
         if (buildSucceed) {
             try {
-                final MarathonBuilder builder = MarathonBuilder.getBuilder(this)
+                MarathonBuilder builder = MarathonBuilder.getBuilder(this)
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 8]
```diff
@@ -182,66 +184,83 @@ public String getAppId() {
         return this.appid;
     }
 
+    @Override
     public String getUrl() {
-        return url;
+        return this.url;
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 9]
```diff
@@ -182,66 +184,83 @@ public String getAppId() {
...
     }
 
     @Override
     public boolean getForceUpdate() {
-        return forceUpdate;
+        return this.forceUpdate;
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 10]
```diff
@@ -182,66 +184,83 @@ public String getAppId() {
...
     }
 
+    @Override
     public String getDocker() {
-        return docker;
+        return this.docker;
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 11]
```diff
@@ -182,66 +184,83 @@ public String getAppId() {
...
+        this.dockerForcePull = dockerForcePull;
+    }
+
+    @Override
     public String getCredentialsId() {
         return this.credentialsId;
     }
 
     @DataBoundSetter
-    public void setCredentialsId(final String credentialsId) {
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 12]
```diff
@@ -182,66 +184,83 @@ public String getAppId() {
...
+    public void setCredentialsId(String credentialsId) {
         this.credentialsId = credentialsId;
     }
 
+    @Override
     public List<MarathonUri> getUris() {
-        return uris;
+        return this.uris;
     }
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 13]
```diff
@@ -182,66 +184,83 @@ public String getAppId() {
...
 
     @DataBoundSetter
-    public void setUris(final List<MarathonUri> uris) {
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java [hunk 14]
```diff
@@ -182,66 +184,83 @@ public String getAppId() {
...
+    public void setUris(List<MarathonUri> uris) {
         this.uris = uris;
     }
 
+    @Override
     public List<MarathonLabel> getLabels() {
-        return labels;
+        return this.labels;
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonRecorder.java]: this seems like an unnecessary change

File: src/main/java/com/mesosphere/velocity/marathon/MarathonStep.java [hunk 15]
```diff
@@ -45,101 +44,108 @@
...
     private       boolean             forceUpdate;
+    private long timeout;
 
     @DataBoundConstructor
-    public MarathonStep(final String url) {
+    public MarathonStep(String url) {
         this.url = MarathonBuilderUtils.rmSlashFromUrl(url);
-        this.uris = new ArrayList<MarathonUri>(5);
-        this.labels = new ArrayList<MarathonLabel>(5);
-        this.env = new ArrayList<MarathonVars>(5);
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/MarathonStep.java]: this seems like an unnecessary change

**Comment by colin-msphere** [comment]
CI failures are due to documentation:

```
marathon/interfaces/AppConfig.java:80: warning: no description for @return
[ERROR] * @return
marathon/impl/MarathonApiImpl.java:32: error: bad use of '>'
[ERROR] * Updating an app. If the timeout set by a value > 0, the method will wait until the deployment is done.
[ERROR] ^
marathon/impl/MarathonApiImpl.java:34: warning: no description for @param
[ERROR] * @param client
```

The `@param` error is repeated for all params in `MarathonApi`.

File: src/main/java/com/mesosphere/velocity/marathon/impl/MarathonApiImpl.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+     */
+    @Override
+    public boolean update(Marathon client, App app, boolean forceUpdate, long timeout) {
+        if (client == null || app == null) {
+            throw new IllegalArgumentException("Necessary parameters are not set!");
+        }
+
+        Result result = client.updateApp(app.getId(), app, forceUpdate);
+        boolean success = false;
+        if (timeout > 0) {
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/impl/MarathonApiImpl.java]: odd behavior... if timeout =< 0 that `success = true` .  I would not expect that timing out == success.

File: src/main/java/com/mesosphere/velocity/marathon/impl/MarathonApiImpl.java [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+        }
+        return success;
+    }
+
+    private boolean waitForDeploymentHasFinished(Marathon client, String deploymentId, long timeout) {
+        LOGGER.info(String.format("Waiting for deployment with id='%s' has finished.", deploymentId != null ? deploymentId : "null"));
+        boolean isActive = true;
+        long startTimestamp = System.currentTimeMillis();
+        while (isActive) {
+            waitForSeconds(5); // 5 seconds
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/impl/MarathonApiImpl.java]: we shouldn't use magic numbers and there is no reason to add a comment for something like this.
and `waitForSeconds` function is really just a wrapper for sleep.  hope about `sleep(DEPLOYMENT_WAIT_FREQ_SECS)`

File: src/main/java/com/mesosphere/velocity/marathon/impl/MarathonApiImpl.java [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+            isActive = client.getDeployments().parallelStream() //
+                    .filter(deployment -> deployment.getId().equals(deploymentId)) //
+                    .findFirst() //
+                    .isPresent();
+            if (timeout > 0 && (System.currentTimeMillis() - startTimestamp) >= timeout) {
+                LOGGER.warning(String.format("The deployment with id='%s' has timed out after %dms.", deploymentId != null ? deploymentId : "null", timeout));
+                break;
```
Comment (kensipe) [inline src/main/java/com/mesosphere/velocity/marathon/impl/MarathonApiImpl.java]: what is the expected behavior of the client for a timeout?   to wait again?   to stop?   to wait for a large timeout value?
I would expect a TimeoutException.

**Review by kensipe — CHANGES_REQUESTED** [review]
@sneufeind it is always great to get community PRs.  Thanks for helping us out!
This PR seems to conflate 1) a desire to change or modify the coding standard and 2) new functionality.   IMO debating formatting will get in the way of pushing the functional changes.    
Could you revert the unnecessary changes needed for this new functionality so we can focus on that?
It will greatly speed up the process.

Thanks!

File: src/main/java/com/mesosphere/velocity/marathon/impl/MarathonApiImpl.java (hunk 17 above)
Comment (sneufeind) [reply src/main/java/com/mesosphere/velocity/marathon/impl/MarathonApiImpl.java]: yep, you're right!

File: src/main/java/com/mesosphere/velocity/marathon/impl/MarathonApiImpl.java (hunk 18 above)
Comment (sneufeind) [reply src/main/java/com/mesosphere/velocity/marathon/impl/MarathonApiImpl.java]: good point! i would also expect a timeout exception. the client should stop and the result should be 'failure'.

**Comment by sneufeind** [comment]
@kensipe @joerg84 @colin-msphere hi all! first of all i really like to thank you for reviewing  this PR and i have to say sorry for the reformatting issues. hopefully, the changes meet your requirements right now. 😄

**Comment by sneufeind** [comment]
hey guys, can you tell me what went wrong in the ci-build?

**Comment by jbomaileva** [comment]
No one is interested by this feature ?

**Comment by mgriffin13** [comment]
Related Jira issue: https://issues.jenkins-ci.org/browse/JENKINS-45699
